include/os/windows/spl/sys/kmem_impl.h

```diff
@@ -0,0 +1,501 @@
+
+/* slab_later_count thresholds */
+#define	KMEM_DISBELIEF		3
+
+/* slab_flags */
+#define	KMEM_SLAB_NOMOVE	0x1
+#define	KMEM_SLAB_MOVE_PENDING	0x2
+
+typedef struct kmem_slab {
+	struct kmem_cache	*slab_cache;	/* controlling cache */
+	void			*slab_base;	/* base of allocated memory */
+	avl_node_t		slab_link;	/* slab linkage */
+	struct kmem_bufctl	*slab_head;	/* first free buffer */
+	long			slab_refcnt;	/* outstanding allocations */
+	long			slab_chunks;	/* chunks (bufs) in this slab */
+	uint32_t		slab_stuck_offset; /* unmoved buffer offset */
+	uint16_t		slab_later_count;  /* cf KMEM_CBRC_LATER */
+	uint16_t		slab_flags;	/* bits to mark the slab */
+	hrtime_t		slab_create_time; /* when was slab created? */
+} kmem_slab_t;
+
+#define	KMEM_HASH_INITIAL	64
+
+#define	KMEM_HASH(cp, buf)	\
+	((cp)->cache_hash_table +	\
+	(((uintptr_t)(buf) >> (cp)->cache_hash_shift) & (cp)->cache_hash_mask))
+
+#define	KMEM_CACHE_NAMELEN 31
+
+typedef struct kmem_magazine {
+	void	*mag_next;
+	void	*mag_round[1];		/* one or more rounds */
+} kmem_magazine_t;
+
+/*
+ * The magazine types for fast per-cpu allocation
+ */
+typedef struct kmem_magtype {
+	short		mt_magsize;	/* magazine size (number of rounds) */
+	int		mt_align;	/* magazine alignment */
+	uint32_t	mt_minbuf;	/* all smaller buffers qualify */
+	uint32_t	mt_maxbuf;	/* no larger buffers qualify */
+	kmem_cache_t	*mt_cache;	/* magazine cache */
+} kmem_magtype_t;
+
+#define	KMEM_CPU_CACHE_SIZE	128	/* must be power of 2 */
+#define	KMEM_CPU_PAD		(KMEM_CPU_CACHE_SIZE - sizeof (kmutex_t) - \
+	2 * sizeof (uint64_t) - 2 * sizeof (void *) - sizeof (int) - \
+	5 * sizeof (short))
+#define	KMEM_CACHE_SIZE(ncpus)	\
+	offsetof(kmem_cache_t, cache_cpu[ncpus])
+
+	/* Offset from kmem_cache->cache_cpu for per cpu caches */
+#define	KMEM_CPU_CACHE_OFFSET(cpuid)	\
+	offsetof(kmem_cache_t, cache_cpu[cpuid]) - \
+	offsetof(kmem_cache_t, cache_cpu)
+
+/*
+ * Per CPU cache data
+ */
+typedef struct kmem_cpu_cache {
+	kmutex_t	cc_lock;	/* protects this cpu's local cache */
+	uint64_t	cc_alloc;	/* allocations from this cpu */
+	uint64_t	cc_free;	/* frees to this cpu */
+	kmem_magazine_t	*cc_loaded;	/* the currently loaded magazine */
+	kmem_magazine_t	*cc_ploaded;	/* the previously loaded magazine */
+	int		cc_flags;	/* CPU-local copy of cache_flags */
+	short		cc_rounds;	/* number of objects in loaded mag */
+	short		cc_prounds;	/* number of objects in previous mag */
+	short		cc_magsize;	/* number of rounds in a full mag */
+	short		cc_dump_rounds;	/* dump time copy of cc_rounds */
+	short		cc_dump_prounds; /* dump time copy of cc_prounds */
+	char		cc_pad[KMEM_CPU_PAD]; /* for nice alignment */
+} kmem_cpu_cache_t;
+
+/*
+ * The magazine lists used in the depot.
+ */
+typedef struct kmem_maglist {
+	kmem_magazine_t	*ml_list;	/* magazine list */
+	long		ml_total;	/* number of magazines */
+	long		ml_min;		/* min since last update */
+	long		ml_reaplimit;	/* max reapable magazines */
```

Comment:
These should be `uint64_t`, rather than `long`, or else we could potentially overflow on really big memory systems. `ml_total` in particular produces the CodeQL warning `Multiplication result may overflow 'long' before it is converted to 'unsigned long long'.`.